### DISCUSSION TITLE
Support for wildcard expanding all available fields

### DISCUSSION THREAD

**Discussion by alanhamlett** [graphql/graphiql#3809] [Q&A / Support] [answered]
Sometimes I would like to see all the data and all fields available, without having to list all of them. In this case, it would be nice if GraphiQL supported me typing `*` and then autocompleted with all available fields. For example, when I type:

```graphql
{
  user {
    *
  }
}
```

If `user` contained `4` fields I would want GraphiQL to autocomplete and fill in the query as:

```graphql
{
  user {
    id
    name
    location
    created_at
  }
}
```

**Comment by alanhamlett** [comment]
Related to https://github.com/graphql/graphql-spec/issues/127, but we don't actually need wildcards in the query protocol... only in the GraphiQL client.

**Comment by acao** [comment]
this used to work in `graphiql` and may be a regression, what happens when you cmd+enter on completion?
https://github.com/graphql/graphiql/blob/4d3eeaa4446c84e92cd77f213e454059602a72e5/packages/graphiql/src/components/GraphiQL.tsx#L997
it may be that this is somehow incompatible with the latest version of codemirror. this is even customizeable via the `getDefaultFieldNames` prop

I started with an experiment of adding this to `monaco-graphql` as well on completion, and had great success. no need for a `*`, a command available as a keyboard shortcut, in the command palette, and as code action (the lightbulb ala vscode) would be easier. 

I'm considering proposing this completion option as well: `fillLeafsOnCompletion`:  "all", "required", or "none".

**Comment by alanhamlett** [comment]
`⌘ + Enter` just sends the query to API for me.

**Comment by itdependsnetworks** [comment]
The `@` symbol seems to work some places but not others, and I am pretty sure I have used other special characters that have works, but again only sometimes and I can't reproduce. 

Working

<img width="460" alt="image" src="https://github.com/user-attachments/assets/22a268e9-7913-4aab-b153-5b32f178060e">

Not Working
<img width="432" alt="image" src="https://github.com/user-attachments/assets/d0472793-6632-400c-a547-fe093017b3d1">

**Comment by itdependsnetworks** [comment]
ok, typing a number seems to work

<img width="503" alt="image" src="https://github.com/user-attachments/assets/a1795dc5-9353-4094-92b1-495a951dd0eb">

**Comment by itdependsnetworks** [comment]
One more... for filters, using a parenthesis seems to work, it has to be empty, so you often have to delete your current filter to get the whole list again. So only when you type `(` does it reliably show up. I have seen `$` and I think others occasionally work, but not consistently. 

![image](https://github.com/user-attachments/assets/ccfefbd8-13a6-4ad5-a31f-8f3074396425)

**Comment by acao** [comment, marked as answer]
I will make a guide to this soon, as it's much easier to auto expand and fill leafs than this, and the getDefaultFieldNames prop callback allows you to endlessly customize this, to the point of allowing you to enable auto-expansion of all fields. it's just a matter of knowing the keyboard shortcuts and how to customize!

1. **shift + space** = expand autocompletion anywhere (also **ctrl+space**, **windows+space** and **cmd+space** works for this)
![image](https://github.com/user-attachments/assets/686c7d8d-8ff3-4b91-a4fe-03ae6ff3f1ca)

2. once a field name is completely typed out, type **cmd/windows + enter** to auto-expand relay-esque defaults as per the [default `getDefaultFieldNames` prop value](https://github.com/graphql/graphiql/blob/cd5f72390578dd37e72c2b40f1c498efa61277ba/packages/graphiql-toolkit/src/graphql-helpers/auto-complete.ts#L86) (note this default callback value may move for 4.x). there is a bug from the 2.x rewrite that causes it to accidentally execute the query here that we need to fix

![image](https://github.com/user-attachments/assets/bfe270cc-4fac-4e5b-82e3-fd67013cda45)

as you can see, if you remove the condition on line 113, you have a recipe for auto-expanding all leafs, something like this maybe:

```tsx

function myCustomGetDefaultFieldNames(type: GraphQLType) {
  // If this type cannot access fields, then return an empty set.
  // if (!type.getFields) {
  if (!('getFields' in type)) {
    return [];
  }

  const fields = type.getFields();

  // Include all fields.
  const leafFieldNames: Array<string> = [];
  for (const fieldName of Object.keys(fields)) {
      leafFieldNames.push(fieldName);
  }
  return leafFieldNames;
}


<GraphiQL getDefaultFieldNames={myCustomGetDefaultFieldNames} />
```

these are all part of the original `GraphiQL` and date back to 0.6.x or earlier iirc.

**Reply by acao** [reply]
@itdependsnetworks see the answer above - no "wildcard" character exists as such in graphql language specification. keyboard shortcuts are much easier

**Reply by itdependsnetworks** [reply]
Shortcuts are great... when you know them. Perhaps there is a way to make the shortcuts a bit more obvious? Like showing them as the first not selectable row in the drop down.

**Reply by acao** [reply]
Agreed! Since ~2016, the documentation for this shortcut appears with new tabs:

![image](https://github.com/user-attachments/assets/82f007c5-253f-4c84-94bc-2d3e98a10773)

we should add it to the readme/future docs site as well
